### PR TITLE
Download testing emulators using a buffer

### PR DESCRIPTION
**Pull request by dlarocque**
The streaming file download was failing when downloading the data connect emulator in CI only (reasons unknown). To fix the issue, we can download the file to a buffer (`Uint8Array`), then write it synchronously to disk.

We preferred the streaming approach because it meant we didn't have to store the entire emulator file in memory, but I found that this buffer approach uses roughly the same amount of memory (~195mB in total for the test process).

### CODE REVIEW

File: scripts/emulator-testing/dataconnect-test-runner.ts [hunk 1]
```diff
@@ -38,7 +39,7 @@ async function run(): Promise<void> {
     await emulator.setUp();
     await runTest(emulator.port);
   } finally {
-    await emulator.tearDown();
```
Comment (DellaBitta) [inline scripts/emulator-testing/dataconnect-test-runner.ts:41]: Why was this change made?

Comment (dlarocque) [reply scripts/emulator-testing/dataconnect-test-runner.ts:41]: It looks like firestore and database also await this expression. We should probably remove await from there too.

**Review by DellaBitta — APPROVED** [review]

File: scripts/emulator-testing/emulators/emulator.ts [hunk 2]
```diff
@@ -53,64 +53,73 @@ export abstract class Emulator {
       return Promise.resolve();
     }
 
+    console.log(`Downloading emulator from [${this.binaryUrl}] ...`);
+    const { name: tempDir } = tmp.dirSync({ unsafeCleanup: true });
+    const filepath = path.resolve(tempDir, this.binaryName);
+    let cur = 0;
+    let buf = new Uint8Array(2 ** 26);
```
[the remark this answers is not in the document]
Comment (dlarocque) [reply scripts/emulator-testing/emulators/emulator.ts]: Done

File: scripts/emulator-testing/emulators/emulator.ts [hunk 3]
```diff
@@ -53,64 +53,78 @@ export abstract class Emulator {
...
-                      writer.write(value);
-                      return reader.read().then(readChunk);
-                    }
-                  });
+      /**
+       * Once the download is `done` in `readChunk`, we want to set `this.binaryPath` to the path of the
+       * downloaded emulator. Unfortunately, we can't access `this` when inside `readChunk`'s scope, since
+       * it's a named function expression, and does not inherit the `this` object from it's parent.
+       * To work around this, we wrap the fetch in a promise,
+       * then once it's resolved we can access `this` in a callback arrow function that *does* inherit
```
Comment (hsubox76) [inline scripts/emulator-testing/emulators/emulator.ts:72]: There's a couple of other ways around this, I think, not 100% sure. One is to set like `const self = this` at the top level (outside of the fetch anyway) and then you call `self.binaryPath` when needed. That looks weird so another better looking option is that you create a function at the top level called `setBinaryPath` that is just like

```
const setBinaryPath = (binaryPath) => this.binaryPath = binaryPath;
```
And then I think you can call setBinaryPath inside readChunk. I'm not 100% sure but I think they should work? If they don't, never mind.

**Pull request closed without merging**